
---
title: "Application : Methylation / Expression"
output: html_document
author: "Hayato Yoshioka (University of Tokyo, AgroParisTech)"
---

# Overview. H.Yoshioka et al. (rrda package)

Application of  Methylation / Expression

# functions
```{r}
library(furrr)
library(dplyr)
library(ggplot2)
library(reshape2)
library(RSpectra)
library(rrda)
library(here)
library(xtable)
library(pheatmap)
library(tictoc)
source("script_rrda/functions_rrda_yoshioka.R")
```

To install breast cancer data from brgedata (Pelegri-Siso et al. 2024)

The data is also available on github https://github.com/Yoska393/rrda/tree/main/RDAdata)

```{ Methylation_Data,eval=FALSE}
library(brgedata)
data(brge_methy)
brge_methy
data(brge_gexp)
brge_gexp

common_samples <- intersect(colnames(brge_gexp), colnames(brge_methy))
brge_gexp_filtered <- brge_gexp[, common_samples]
brge_methy_filtered <- brge_methy[, common_samples]

methy <- t(assay(brge_methy_filtered, "Beta"))
gexp <- t(as.matrix(brge_gexp_filtered))

mg<-list(
	methy=methy,
	gexp=gexp
)

saveRDS(mg,here(RDAdata,Methdata.RDS))
```


## data
```{r}
base_path <- here("out", "Meth")

if (!dir.exists(base_path)) {
  dir.create(base_path, recursive = TRUE)
  message("Directory created: ", base_path)
} 

MethData<-readRDS(here("RDAdata","Methdata.RDS"))
gexp <- MethData$gexp
methy <- MethData$methy

Y <- gexp
X <- methy

rn<-1

X_train<-X[-rn,]
X_test<-X[rn,,drop=F]
Y_train<-Y[-rn,]
Y_test<-Y[rn,,drop=F]

dim(Y_train)
dim(Y_test)
dim(X_train)
dim(X_test)

```

## One result
```{R,eval=F}
set.seed(123)
n_lam <- 50

Lambda <- exp(seq(2, 13, length.out = n_lam))

plan(multisession)
cv_result<- rrda.cv(Y = Y_train, X = X_train, lambda= Lambda,maxrank =15,nfold = 5,scale.X = F,scale.Y = F)
plan(sequential)

rrda.summary(cv_result)
saveRDS(cv_result,file.path(base_path,"path.RDS"))
```

```{r,warning=FALSE,fig.width=9,fig.height=6}
cv_result<-readRDS(file.path(base_path,"path.RDS"))
rrda.plot(cv_result = cv_result,show_error_bar = TRUE)
rrda.plot(cv_result = cv_result,show_error_bar = FALSE)

rrda.heatmap(cv_result,nrank = c(1:20))
K<-cv_result$opt_min$rank
L<-cv_result$opt_min$lambda

```

```{r,fig.width=6,fig.height=6}

best_Bhat <- rrda.fit(Y = Y_train, X = X_train, nrank = K, lambda = L,scale.X = F,scale.Y = F)
best_Pred <- rrda.predict(Bhat = best_Bhat, X = X_test)[[1]][[1]][[1]]

lim <- range(c(Y_test, best_Pred)) 
plot(Y_test, best_Pred, cex=0.1,cex.lab=1.3, xlim=lim, ylim=lim,xlab="Gene Expression",ylab="Predicted Gene Expression")
abline(0, 1, col=2)
```


# session info
check the info and environment
```{r}
devtools::session_info()
```
