---
title: "Application : Methylation / Expression"
output: html_document
author: "Hayato Yoshioka (University of Tokyo, AgroParisTech)"
---

# Overview. H.Yoshioka et al. (rrda package)

Application of Methylation / Expression

# functions
```{r}
library(furrr)
library(dplyr)
library(ggplot2)
library(reshape2)
library(RSpectra)
library(rrda)
library(here)
library(xtable)
library(pheatmap)
library(tictoc)
source("script_rrda/functions_rrda_yoshioka.R")

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19")

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)

```

To install breast cancer data from brgedata (Pelegri-Siso et al. 2024)

The data is also available on github https://github.com/Yoska393/rrda/tree/main/RDAdata)

```{ Methylation_Data,eval=FALSE}
library(brgedata)
data(brge_methy)
brge_methy
data(brge_gexp)
brge_gexp

common_samples <- intersect(colnames(brge_gexp), colnames(brge_methy))
brge_gexp_filtered <- brge_gexp[, common_samples]
brge_methy_filtered <- brge_methy[, common_samples]

methy <- t(assay(brge_methy_filtered, "Beta"))
gexp <- t(as.matrix(brge_gexp_filtered))

mg<-list(
	methy=methy,
	gexp=gexp
)

saveRDS(mg,here(RDAdata,Methdata.RDS))
```


## data
```{r}
base_path <- here("out", "Meth")

if (!dir.exists(base_path)) {
  dir.create(base_path, recursive = TRUE)
  message("Directory created: ", base_path)
} 

MethData<-readRDS(here("RDAdata","Methdata.RDS"))
gexp <- MethData$gexp
methy <- MethData$methy

Y <- gexp
X <- methy

rn<-1

X_train<-X[-rn,]
X_test<-X[rn,,drop=F]
Y_train<-Y[-rn,]
Y_test<-Y[rn,,drop=F]

dim(Y_train)
dim(Y_test)
dim(X_train)
dim(X_test)

```

## One result
```{R,eval=F}
set.seed(123)
n_lam <- 50

Lambda <- exp(seq(2, 13, length.out = n_lam))

plan(multisession)
cv_result<- rrda.cv(Y = Y_train, X = X_train, lambda= Lambda,maxrank =15,nfold = 5,scale.X = F,scale.Y = F)
plan(sequential)

rrda.summary(cv_result)
saveRDS(cv_result,file.path(base_path,"path.RDS"))
```

```{r,warning=FALSE,fig.width=9,fig.height=6}
cv_result<-readRDS(file.path(base_path,"path.RDS"))
rrda.plot(cv_result = cv_result,show_error_bar = TRUE)
rrda.plot(cv_result = cv_result,show_error_bar = FALSE)

rrda.heatmap(cv_result,nrank = c(1:20))
K<-cv_result$opt_min$rank
L<-cv_result$opt_min$lambda

```

```{r,fig.width=6,fig.height=6}

best_Bhat <- rrda.fit(Y = Y_train, X = X_train, nrank = K, lambda = L,scale.X = F,scale.Y = F)
best_Pred <- rrda.predict(Bhat = best_Bhat, X = X_test)[[1]][[1]][[1]]

lim <- range(c(Y_test, best_Pred)) 
plot(Y_test, best_Pred, cex=0.1,cex.lab=1.3, xlim=lim, ylim=lim,xlab="Gene Expression",ylab="Predicted Gene Expression")
abline(0, 1, col=2)
```


## All
```{R,eval=F}
set.seed(123)
n_lam <- 50

Lambda <- exp(seq(2, 13, length.out = n_lam))

plan(multisession)
cv_result<- rrda.cv(Y = Y, X = X, lambda= Lambda,maxrank =15,nfold = 5,scale.X = F,scale.Y = F)
plan(sequential)

rrda.summary(cv_result)
saveRDS(cv_result,file.path(base_path,"path_all.RDS"))
```

```{r,warning=FALSE,fig.width=9,fig.height=6}
cv_result<-readRDS(file.path(base_path,"path_all.RDS"))
rrda.plot(cv_result = cv_result,show_error_bar = TRUE)
rrda.plot(cv_result = cv_result,show_error_bar = FALSE)

rrda.heatmap(cv_result,nrank = c(1:20))
```

```{r,warning=FALSE}
# Extract optimal lambda and rank
L <- cv_result$opt_min$lambda
K <- cv_result$opt_min$rank

# Fit the model using the optimal lambda and rank
b <- rrda.fit(Y = Y, X = X, lambda = L, nrank = K)

b1<-b$Bhat_comp[[1]][[1]]
b2<-b$Bhat_comp[[1]][[2]]

dim(b1)
dim(b2)

rownames(b1)<-colnames(X)
rownames(b2)<-colnames(Y)

x_scores <- apply(b1, 1, function(row) sqrt(sum(row^2)))
top_x_idx <- order(x_scores, decreasing = TRUE)[1:30]  

y_scores <- apply(b2, 1, function(row) sqrt(sum(row^2)))
top_y_idx <- order(y_scores, decreasing = TRUE)[1:30]

b1_sub <- b1[top_x_idx, , drop = FALSE]  
b2_sub <- b2[top_y_idx, , drop = FALSE] 

B_sub <- b1_sub %*% t(b2_sub)
dim(B_sub)
```


```{r,warning=FALSE,fig.width=9,fig.height=7}

unc <- !grepl("^TCUn", colnames(B_sub))
filtered_matrix<-B_sub[,unc]

dim(filtered_matrix)
h<-filtered_matrix

max_value <- max(h, na.rm = TRUE)
min_value <- min(h, na.rm = TRUE)

custom_colors <- colorRampPalette(c("blue", "white", "red"))(200)
max_abs_value <- max(abs(c(min_value, max_value)))
breaks <- seq(-max_abs_value, max_abs_value, length.out = 201)

# 
# 
# pheatmap(h,
# 				 color = custom_colors,
# 				 breaks = breaks,
# 				 border_color = NA,
# 				 #scale = "row",
# 				 clustering_distance_rows = "euclidean",
# 				 clustering_distance_cols = "euclidean",
# 				 clustering_method = "complete",
# 				 show_rownames = TRUE,
# 				 show_colnames = TRUE,
# 				 xlab = "Methylation",
# 				 ylab = "Expression",
# 				 fontsize_row = 8,
# 				 fontsize_col =8,
# 				 width = 10,
# 				 height = 15
# )


# annotation
anno <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)

cg_ids <- rownames(h) 

anno_subset <- anno[rownames(anno) %in% cg_ids, ]


result <- data.frame(
  cg_id = rownames(anno_subset),
  chr = anno_subset$chr,
  pos = anno_subset$pos,
  gene = anno_subset$UCSC_RefGene_Name,
  relation_to_cpg = anno_subset$Relation_to_Island,
  stringsAsFactors = FALSE
)


# 1. Get current CpG IDs from rownames
cg_ids <- rownames(h)

# 2. Subset annotation data for matching CpG IDs
anno_subset <- anno[rownames(anno) %in% cg_ids, ]

# 3. Reorder annotation to match the order in 'h'
anno_subset <- anno_subset[cg_ids[cg_ids %in% rownames(anno_subset)], ]

# 4. Create new row names with chromosome name *after* cg ID
new_rownames <- paste0(anno_subset$chr, "_",rownames(anno_subset) )

# 5. Replace row names in 'h' (only for matching CpG IDs)
rownames(h)[rownames(h) %in% rownames(anno_subset)] <- new_rownames
# 6. Get original column names
# 1. Get the original column names
original_colnames <- colnames(h)

# 2. Extract the two characters immediately after "TC"
chr_code <- substr(original_colnames, 3, 4)  # 3rd and 4th characters = chromosome code

# 3. Create chr label (e.g., "chr02", "chrX", "chr13")
chr_labels <- ifelse(chr_code %in% c("0X", "0Y"),
                     paste0("chr", substring(chr_code, 2)),  # "0X" â†’ "chrX"
                     paste0("chr", chr_code))                # keep "02", "13" as-is

# 4. Combine with original column names
new_colnames <- paste0(chr_labels, "_", original_colnames)

# 5. Keep only valid chromosome labels (skip NA or malformed)
keep_cols <- !grepl("^chrNA", chr_labels)
h <- h[, keep_cols]
colnames(h) <- new_colnames[keep_cols]

pheatmap(h,
				 color = custom_colors,
				 breaks = breaks,
				 border_color = NA,
				 #scale = "row",
				 clustering_distance_rows = "euclidean",
				 clustering_distance_cols = "euclidean",
				 clustering_method = "complete",
				 show_rownames = TRUE,
				 show_colnames = TRUE,
				 xlab = "Methylation",
				 ylab = "Expression",
				 fontsize_row = 8,
				 fontsize_col =8,
				 width = 10,
				 height = 15
)
```




# session info
check the info and environment
```{r}
devtools::session_info()
```
